### PR TITLE
Make enum addValue() faster by using an Map rather than an array

### PR DESCRIPTION
**Pull request by clenfest**
Merging enums was previously a O(n^3) operation.

### CODE REVIEW

File: internals-js/src/utils.ts [hunk 1]
```diff
@@ -479,3 +479,29 @@ export function setsEqual<T>(s1: Set<T> | null, s2: Set<T> | null): boolean {
   }
   return true;
 }
+
+// ordering is done based on order that elements were added
+export class TimeOrderedSet<V extends { name: string }> {
+  private _values: V[] = [];
+  private _valueMap: Map<string, V> = new Map();
```
Comment (benjamn) [inline internals-js/src/utils.ts:486]: In ECMAScript/JS, `Map` and `Set` are specified to preserve insertion order when you iterate over them, so you might not need this wrapper class?

File: internals-js/src/definitions.ts [hunk 2]
```diff
@@ -2319,7 +2319,7 @@ export class EnumType extends BaseNamedType<OutputTypeReferencer, EnumType> {
     }
     const existing = this.value(toAdd.name);
     if (!existing) {
-      this._values.push(toAdd);
+      this._values.add(toAdd);
```
Comment (benjamn) [inline internals-js/src/definitions.ts:2322]: An important question with `Map` and `Set` is what you want to happen when you add an element that's already in the container. By default, it won't be bumped up to the most recent / last position, but you can achieve that behavior by deleting the item before adding it (if it already exists).

However, it looks like this code only ever calls `this._values.add` when the name does not yet exist, so using a `Map` for `this._values` should work without the trick of deleting elements.

**Review by benjamn — CHANGES_REQUESTED** [review]
I'd like to see if this can be simplified to use a `Map` with no wrapper class:
```ts
private _values = new Map<string, EnumValue>();
```

File: internals-js/src/utils.ts (hunk 1 above)
Comment (clenfest) [reply internals-js/src/utils.ts:486]: Interesting. I didn't realize that. I'll change it.

File: internals-js/src/definitions.ts (hunk 2 above)
Comment (clenfest) [reply internals-js/src/definitions.ts:2322]: I think the default behavior should be fine.

**Review by benjamn — APPROVED** [review]

**Pull request merged**